[TEP-0091] Update trusted resources feature flag and add condition

Yongxuanzhang opened the pull request:
This commit proposes to 

1. add  `mode` field into`VerificationPolicy`, update the feature flag to `verification-no-match-policy` with allow / warn / deny   
2. add condition into taskrun and pipelinerun to indicate the verification passes or not.

Yongxuanzhang commented:
Hi @afrittoli @wlynch @bobcatfish @jagathprakash 
This is a proposal to update the trusted resources tep, plz take a look and share your comments. 
Thanks!

jerop commented:
/kind tep

jagathprakash commented on teps/0091-trusted-resources.md:
```diff
@@ -377,13 +377,18 @@ metadata:
- * `skip`: Directly skip the verification.
+`resource-verification-mode`. (Optional, `verify_all`, `verify_match` or `no_verification`, default `no_verification`):
+ * `verify_all`: Failing verification will mark the taskruns/pipelineruns as failed.
+ * `verify_match`: fetch VerificationPolicy from the namespace, and if policies matched then verify, otherwise skip the verification.
```
nit:
s/fetch/Fetch

Yongxuanzhang replied:
thx!

jagathprakash commented on teps/0091-trusted-resources.md:
```diff
@@ -377,13 +377,18 @@ metadata:
- * `skip`: Directly skip the verification.
+`resource-verification-mode`. (Optional, `verify_all`, `verify_match` or `no_verification`, default `no_verification`):
+ * `verify_all`: Failing verification will mark the taskruns/pipelineruns as failed.
+ * `verify_match`: fetch VerificationPolicy from the namespace, and if policies matched then verify, otherwise skip the verification.
```
nit: This could be explained elsewhere, but by match it is not clear what is supposed to match. Maybe a link to matching point 2 in "Verify Resources" could be useful.

Yongxuanzhang replied:
👍

pritidesai commented:
API WG - waiting for the review

/assign @jagathprakash 
/assign @afrittoli

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -377,13 +377,18 @@ metadata:
- * `skip`: Directly skip the verification.
+`resource-verification-mode`. (Optional, `verify_all`, `verify_match` or `no_verification`, default `no_verification`):
+ * `verify_all`: Failing verification will mark the taskruns/pipelineruns as failed.
+ * `verify_match`: Fetch VerificationPolicy from the namespace, and if policies matched then verify, otherwise skip the verification. Details explained in [ Verify the Resources](#verify-the-resources)
```
I don't think this is a good idea to support globally, though I get why we'd want something like this for incremental adoption. It would be generally be bad if Tasks could circumvent verification just by renaming to something that's not covered by policy, and new Tasks that are probably riskier would be exempted by default since they're unlikely to match existing policies.

Instead of making verification opt-in, it would probably be safer to make the mode opt-out on a per-policy basis (see [`ClusterImagePolicySpec.mode`](https://github.com/sigstore/policy-controller/blob/main/docs/api-types/index.md#clusterimagepolicyspec) for an example). This way things are directed to the more secure way by default, forcing users to either configure the matching policy or explicitly opt out the Tasks.

Yongxuanzhang replied:
Thanks! That makes sense. By looking at the [ClusterImagePolicySpec.mode](https://github.com/sigstore/policy-controller/blob/main/docs/api-types/index.md#clusterimagepolicyspec).  How about changing the `verify_match` to:
1. If found matched polices then verify and fail
2. If no matched policies then warn

wlynch replied:
Preferring warn over skip generally sounds good, but that doesn't address the larger problem of opt-in enforcement and the ability to evade enforcement by crafting a non-matching artifact. 😢 

Having policies that opt-out matching patterns seems like a better global policy - this way new artifacts that we haven't encountered before (whether they are malicious or not) can't bypass enforcement.

wlynch commented:
Followed up with @Yongxuanzhang off thread - 

We're leaning towards modifying this slightly by:

1. adding a `mode` field to `VerificationPolicy` to control allow/warn behavior rather than having a global config map value.
2. Modifying `resources-verification-mode` to specify the action to take if no policy matches (we may rename this to make the behavior clearer).

If you want to have the same affect as what is proposed for `verify-match`, you can achieve this in one of 2 ways:

1. Set ConfigMap `resources-verification-mode: enforce`, then define a policy like:

    ```yaml
    apiVersion: tekton.dev/v1alpha1
    kind: VerificationPolicy
    metadata:
      name: foo
    spec:
      resources:
        - pattern: ".*"
      mode: warn
    ```

    Ideally this allowlist should be scoped to only a few sets of resources, and acts as an allowlist. Ideally for cases like "we want to enforce everything, but we know a set of resources isn't signed yet so we will allow those". Enforce mode policies would take precedence over warn.

2. Set ConfigMap `resources-verification-mode: warn`, then define a policy like:

    ```yaml
    apiVersion: tekton.dev/v1alpha1
    kind: VerificationPolicy
    metadata:
      name: foo
    spec:
      resources:
        - pattern: "https://github.com/tektoncd/catalog.git"
      mode: enforce  # this would be the default so could also be omitted.
    ```
  
    This would be closer to what was originally proposed for `resources-verification-mode` - i.e. only enforce matching policies without an explicit opt-out policy.

Much of this was inspired by https://github.com/sigstore/policy-controller so, feel free to take a look there to read up on additional background.

We'll plan to discuss this more in next week's s3c working group, so please join if you have thoughts / opinions! 🙏

Yongxuanzhang commented:
/assign @bobcatfish

Yongxuanzhang commented:
/assign @jagathprakash

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -377,13 +386,44 @@ metadata:
- * `warn`: Log warning but don't fail the taskruns/pipelineruns.
- * `skip`: Directly skip the verification.
+`enable-trusted-resource`. (Optional, `true` or `false`, default `false`):
+ * `true`: Failing verification will mark the taskruns/pipelineruns as failed.
```
Warn throws a wrench into this. This might be another reason to just drop the feature flag entirely and just rely on policy existence. 🤔

Yongxuanzhang replied:
I was thinking about implementation details and thought having this might be easier. 

Without this, if we want to turn off this feature:
we need to set the no-match-policy to "allow", and make sure no policies are in the namespace. This may need some operations on the cluster. But if we have the feature flag, users can just turn off there without removing the policies. 

I think another difference between this work and policy-controller is that, trusted resources is just one feature of tekton pipeline. Policy controller is a project designed for images verification. Users will use it once they deploy the pods, and if they don't want to use they can remote the deployment.

Yongxuanzhang replied:
updated to remove this and rename the current feature flag to "no-match-policy"

wlynch commented on teps/0091-trusted-resources.md:
````diff
@@ -377,13 +386,44 @@ metadata:
+  no-match-policy: "deny"
+```
+
+`no-match-policy`. (Optional, `allow`, `deny` or `warn`, default `deny`):
````
You probably want to default to allow, at least while still in alpha - this is the other piece that will let you drop the 
feature flag.

This may end up being a feature we always start at allow/warn - we probably don't want to flip on enforcement until a user is ready.

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -377,13 +386,44 @@ metadata:
+ * `allow`: when no matching policies found, allow the resource.
+ * `warn`: when no matching policies found, allow the resource and log warning.
+
+ **Note:** If users want to have both signed resources and unsigned resources. The `enable-trusted-resource` needs to be set to `true` and `no-match-policy` is set to `allow`. So for signed resources, matched policies can be used to verify. For unsigned resources, no policies will be found so they are allowed.
```
This isn't quite true, since you could also have warn. What might be more useful here are some examples of how the policies + flag interact with each other (this would also be a good chance to document how multiple matching policies interact)

wlynch commented on teps/0091-trusted-resources.md:
````diff
@@ -377,13 +386,44 @@ metadata:
+
+ #### Condition Update
+
+ Trusted resources should also update the taskrun/pipelinerun’s condition to indicate if it passes verification or not. This can be done via knative api.
+
+ Condition of a failure should be:
+ ```go
+ apis.Condition{
+			Type:    apis.ConditionSucceeded,
+			Status:  corev1.ConditionFalse,
+			Reason:  podconvert.ReasonResourceVerificationFailed,
+			Message: "Trusted resource verification failed",
+		}
+ ```
````
You probably want to go more in depth here, since there are actually different states you can be in that will do different things:

1. Pass verification
2. Fail verification, enforce
3. Fail verification, warn
4. Fail verification, allow

tables / diagrams might help here.

jagathprakash replied:
+1, A flow chart or table would help.

jagathprakash commented on teps/0091-trusted-resources.md:
````diff
@@ -377,13 +386,44 @@ metadata:
+  no-match-policy: "deny"
+```
+
+`no-match-policy`. (Optional, `allow`, `deny` or `warn`, default `deny`):
````
Maybe a flow chart is useful here.
I presume if enable-trusted-resource is false then entire trusted resource code is not touched, so no messaging to the user.
If true and no-match-policy is allow then resources without policies will be allow, but will there be a message that it was allowed? Will a condition be set here? How does user know that the resource did not have a verification policy but was still allowed? Maybe another Reason in conidition?
If true and no-match-policy is warn then resources without policies will be warned. "Warn" is defined as set the condition Reason to "podconvert.ReasonResourceVerificationFailed," ?

Yongxuanzhang replied:
Thanks!

jagathprakash commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +373,70 @@ type KeylessRef struct {
- * `skip`: Directly skip the verification.
+`no-match-policy`. (Optional, `allow`, `deny` or `warn`, default `allow`):
+ * `deny`: Reject the resource and fail the taskrun/pipelinerun if no matching policies are found.
+ * `allow`: Allow the resource if no matching policies are found.
```
Since warn is a no-op as far as execution goes, do we need allow? Cant it be just deny or warn?

Yongxuanzhang replied:
I think we need? This is for the feature flag not the policy mode

bobcatfish commented on teps/0091-trusted-resources.md:
```diff
@@ -312,6 +316,11 @@ type VerificationPolicySpec struct {
+	// enforce - Reject (default)
+	// warn - allow but warn
+	// +optional
+	Mode string `json:"mode,omitempty"`
```
nit: i'm confused by the text a few lines up "(Reference from [policy-controller](https://github.com/sigstore/policy-controller))" - is that meaning to say that this code is copied directly from the policy-controller as an example of what the API could be, or is it saying that the code here is inspired by the policy-controller?

If it's the latter (and i think it is?) it might be clearer to say something like "inspired by the policy-controller" or "based on the policy-controller" or have a separate section with something like "similar approaches in other projects"

Yongxuanzhang replied:
Thanks! It is the latter. Will update

Yongxuanzhang replied:
Updated it and also the docstrings.

bobcatfish commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +373,70 @@ type KeylessRef struct {
   namespace: tekton-pipelines
 data:
-  verification-policy: enforce
+  no-match-policy: "deny"
```
naming nit: can we use something that explains a bit more context than 'no-match-policy'? ideas:

- 'verification-policy-unmatched'
- 'verification-no-match-policy'

I think I actually like the original name 'verification-policy' a bit better; instead of removing it you could update the options to be something like

- `require-policy` (or `deny-unmatched`): require that a policy be set for all resources (same as no-match-policy: deny)
- `allow-no-policy` (or `allow-unmatched`?): default (same as no-match-policy: allow)

Yongxuanzhang replied:
Sure! I like the "update the options" approach. Will update this part

Yongxuanzhang replied:
Updated to use 'verification-policy'  with 

- allow-no-policy
- allow-no-policy-warn
- require-policy

wlynch replied:
To me these new values are more confusing - they're less clear on what the resulting behavior is at a glance.

e.g. `allow-no-policy-warn` could be interpreted as "warn if no policy" or also as "don't allow warn policies". Together you can piece together the differences, but I worry this will be more difficult to determine if looking at only a single value in the config map. I think the `no-match-policy` key better matches the condition we're checking, so I'd prefer to stick with that.

No objections to `verification-policy-unmatched` / `verification-no-match-policy` - can we change back to this with the allow / warn / deny values?

Yongxuanzhang replied:
ah ok, I will change the naming to "verification-no-match-policy" with allow / warn / deny values!

Yongxuanzhang replied:
Updated!

bobcatfish commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +373,70 @@ type KeylessRef struct {
+3) Skip the verification if no policies exist and `no-match-policy` is set to `allow`.
+
+ **Note:**
+ * If the `no-match-policy` is set to `deny`, resources that do not match any policies will be rejected. Resources that match policies will be accepted if all the matching `enforce` policies pass or only match `warn` policies.
```
when you say "rejected" do you mean they can't be written to the cluster (enforced by a webhook when applying a Task/Pipeline), or that TaskRuns/PipelineRuns that use them will fail?

Yongxuanzhang replied:
Thanks! I think I need to clear about the wording and make clarifications. 
I think ideally it should mean both. If a task/pipeline fails verification, it should not be written into the cluster and TaskRuns/PipelineRuns that use them will fail. 

For remote resources, they are resolved by remote resolution then verified by trusted resources and will not be written into storage. 
For api resources, we can add verification at webhook to reject them

Yongxuanzhang replied:
Updated related words

bobcatfish commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +373,70 @@ type KeylessRef struct {
+			Reason:  podconvert.ReasonResourceVerificationFailed
+			Message: "Trusted resource verification failed",
+		}
+ ```
````
can you provide an example of what this looks like in the resulting yaml status? e.g. for [execution status of a TaskRun](https://github.com/tektoncd/pipeline/blob/main/docs/taskruns.md#monitoring-execution-status):

```yaml
status:
  completionTime: "2022-12-09T23:39:53Z"
  conditions:
  - lastTransitionTime: "2022-12-09T23:39:53Z"
    message: All Steps have completed executing
    reason: Succeeded
    status: "True"
    type: Succeeded
```

Yongxuanzhang replied:
Yeah sure! I will also explain how our condition co-exist with current conditions!

bobcatfish commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +373,70 @@ type KeylessRef struct {
+Sample Condition:
+ ```go
+ apis.Condition{
+			Type:    apis.ConditionType(apis.ConditionSeverityInfo),
````
I'm a bit confused by this Type, my understanding for our existing execution status is that we use apis.ConditionSucceeded, what is apis.ConditionSeverityInfo intended to be used for, and do we want to use that or do we want to use something custom? (would be useful to provide more detail around how we chose this type and what other options would be)

Yongxuanzhang replied:
Sure! If we use `apis.ConditionSucceeded` then it will be overwritten when taskrun is completed, using a different type can avoid this. The type in this example is from knative:
https://github.com/knative/pkg/blob/980a33719a10024f45e44320316c5bd35cef18d6/apis/condition_types.go#L52

The condition in taskrun status will be:
```
  - lastTransitionTime: "2023-03-01T18:17:05Z"
    message: Trusted resource verification passed
    reason: Verification Success
    status: "True"
    type: Info
```

I didn't think about creating our own condition type before but that could be another option. If we have a custom type then we could also avoid overwritten of the same type

bobcatfish reviewed (commented):
The general approach seems reasonable to me! A few requets:
* I think we could probably pick a config option that is a bit more self explanatory 
* I think there might be a bit more detail to provide around the Condition update, for example are there any alternatives and what are the pros/cons? (e.g. using an annotation instead) It would also be useful to explain how this new condition would co-exist with existing condition(s) on TaskRuns/PipelineRuns
* Could you include a description of how the new config option would be phased in? e.g. would you support the old config option and the new for one release, if so how would they work together and if not, we should call out that we plan to make a backwards incompatible change within one release

jerop commented:
/assign

Yongxuanzhang commented:
> The general approach seems reasonable to me! A few requets:
> 
> * I think we could probably pick a config option that is a bit more self explanatory
> * I think there might be a bit more detail to provide around the Condition update, for example are there any alternatives and what are the pros/cons? (e.g. using an annotation instead) It would also be useful to explain how this new condition would co-exist with existing condition(s) on TaskRuns/PipelineRuns
> * Could you include a description of how the new config option would be phased in? e.g. would you support the old config option and the new for one release, if so how would they work together and if not, we should call out that we plan to make a backwards incompatible change within one release

Thanks for all the suggestions! 

I have updated the PR to include several changes:

1. update the config options to be more explanatory
2. Add alternatives sections with pros and cons, add how this new condition co-exist with existing ones
3. Add a call out that the new config change is backward incompatible
4. Addressed comments

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +373,140 @@ type KeylessRef struct {
+ ```go
+ apis.Condition{
+			Type:    ConditionVerificationFail,
+			Status:  corev1.ConditionFalse,
````
Oh I see, thanks!

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +373,138 @@ type KeylessRef struct {
+ Need to create conditions for trusted resources so this can co-exist with current `apis.ConditionSucceeded` type:
+ ```go
+ const (
+	// ConditionVerification specifies that the resources pass trusted resources verification or not.
+	ConditionVerification apis.ConditionType = "TrustedResourcesVerification"
````
It looks like knative uses `ConditionXXX` as a naming pattern for conditions

wlynch reviewed (commented):
/lgtm

@bobcatfish PTAL if you have any objections (I think most of my recent feedback was in response to your comments).

I can merge this in EOW if there's no other feedback.

bobcatfish commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +373,110 @@ type KeylessRef struct {
+2) Not return error, but warn the log and update the status condition with `FailedVerification` reason if only found `warn` policies.
+3) Skip the verification if no policies exist and `verification-no-match-policy` is set to `allow`.
+
+**Note:** The current proposed `verification-no-match-policy` will replace the old `resource-verification-mode` in one release and this is not a backwards compatible change.
```
Could you elaborate a bit more on whether both options will exist for one release or whether `resource-verification-mode` will be removed and `verification-no-match-policy` will be added in the same release? The latter afaik is a backwards incompatible change. The former isn't but it would be good to understand how the options will coexist for that one release

Yongxuanzhang replied:
We decided with the latter one and I have added the note here. I will update the notice to make it more clear.

bobcatfish commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +373,110 @@ type KeylessRef struct {
+    message: All Steps have completed executing
+    reason: Succeeded
+    status: "True"
+    type: Succeeded
```
I think I'm revisiting a previous discussion, but if the Task fails verification, shouldn't we fail execution? I can't imagine a scenario where a Task fails verification but we want it to execute anyway

Yongxuanzhang replied:
Oh that's true, I think for warn mode policy, we shouldn't fail the execution, for enforce mode we should fail. I will update this part

bobcatfish reviewed (commented):
Looking good overall, I've gotten a bit stuck on the behavior when a Task fails verification. I apologize if I'm pulling in more scope with that discussion 😬

jerop reviewed (commented):
Thanks @Yongxuanzhang!

Left some questions and ideas to clarify the proposal

jerop commented:
already has an lgtm so holding so it's not merged accidentally on the first approval

/hold

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,127 @@ type KeylessRef struct {
+ Need to create conditions for trusted resources so this can co-exist with current `apis.ConditionSucceeded` type:
+ ```go
+ const (
+	// ConditionVerification specifies that the resources pass trusted resources verification or not.
+	ConditionVerification apis.ConditionType = "TrustedResourcesVerification"
````
sure! sgtm!

jerop replied:
found [conditions conventions](https://github.com/kubernetes/community/blob/4c9ef2d/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) that back this request to rename the condition:

> Condition type names should describe the current observed state of the
  resource, rather than describing the current state transitions. This
  typically means that the name should be an adjective ("Ready", "OutOfDisk")
  or a past-tense verb ("Succeeded", "Failed") rather than a present-tense verb
  ("Deploying"). Intermediate states may be indicated by setting the status of
  the condition to `Unknown`.

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,127 @@ type KeylessRef struct {
+Condition:
+ ```go
+ apis.Condition{
+			Type:    ConditionVerification,
````
That makes sense! Thanks!

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +372,127 @@ type KeylessRef struct {
- * `enforce`: Failing verification will mark the taskruns/pipelineruns as failed.
- * `warn`: Log warning but don't fail the taskruns/pipelineruns.
- * `skip`: Directly skip the verification.
+`verification-no-match-policy`. (Optional, `allow`, `warn` or `deny`, default to `allow`):
+ * `allow`: Don't fail the taskrun/pipelinerun if no matching policies are found. Don't log.
+ * `warn`: Don't fail the taskrun/pipelinerun and log a warning if no matching policies are found.
+ * `deny`: Fail the taskrun/pipelinerun if no matching policies are found.
+
+ Reconciler will respond to verification failure based on ConfigMap and Verification Policy exsitence settings to:
+1) Return error in reconciler, mark the status as error with `FailedVerification` reason if `enforce` policies fail
+2) Not return error, but warn the log and update the status condition with `FailedVerification` reason if only found `warn` policies.
+3) Skip the verification if no policies exist and `verification-no-match-policy` is set to `allow`.
```
I found this part overlaps with section below. I have merged them and added examples

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+  ```
+
+   * Matching Policies found:
+     * Don't fail taskrun/pipelinerun if all the matched `enforce` policies pass or only match `warn` policies.
````
this contradicts line 461 which says that warn policies failing causes `Conditions.TrustedResourcesVerified` to be set to "False"

jerop replied:
```suggestion
     * If all the matched `enforce` policies pass, and the TaskRun/PipelineRun succeeds:
```

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+      ```
+
+ * If the `trusted-resources-verification-no-match-policy` is set to `ignore` or `warn`,
+   * No Matching Policies found: Don't fail taskrun/pipelinerun, skip the verification if set to `ignore`, and emit warning log if set to `warn`.
````
so there's no `TrustedResourcesVerified` in this case? if so, please make it explicit

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+          status: "True"
+          type: Succeeded
+      ```
+   * Matching Policies found: Same as `fail`.
````
what does "same as `fail` mean"? it'd be helpful if you restate what happens, or reference the section that it's the same as

jerop reviewed (commented):
does "log warning" mean the same thing as "setting "Conditions.TrustedResourcesVerified" to false? if not, let's clarify that because it seems to be used interchangeably

jerop reviewed (commented):
I'm still confused about the "conditions" as its currently written, but tried to summarize the different combinations in two tables. Is the summary below accurate? If so, can we use these tables to clarify the design?

----

### No Matching Policies

|                             | `Conditions.TrustedResourcesVerified` | `Conditions.Succeeded` |
|-----------------------------|---------------------------------------|------------------------|
| `no-match-policy`: "ignore" |                                       |                        |
| `no-match-policy`: "warn"   | False                                 |                        |
| `no-match-policy`: "fail"   | False                                 | False                  |

### Matching Policies

|                          | `Conditions.TrustedResourcesVerified` | `Conditions.Succeeded` |
|--------------------------|---------------------------------------|------------------------|
| any enforce policy fails | False                                 | False                  |
| only warn policies fail  | False                                 |                        |

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+  ```
+
+   * Matching Policies found:
+     * Don't fail taskrun/pipelinerun if all the matched `enforce` policies pass or only match `warn` policies.
````
if only match `warn` policies, and they fail verification, it won't fail the run and set the `Succeeded ` to false. This is aligned with the table. Why it contradicts 461 sample?

jerop replied:
reading it again, it's not contradictory, but it's confusing when the two scenarios are combined (that's why I thought it was contradicting)

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -312,6 +316,10 @@ type VerificationPolicySpec struct {
 	Authorities []Authority `json:"authorities"`
+	// Mode controls whether a failing policy will fail the taskrun/pipelinerun, or only log the warnings
+	// enforce - fail the taskrun/pipelinerun if verification fails (default)
+	// warn - don't fail the taskrun/pipelinerun if verification fails but log warnings
```
Thanks! This is much better!

jerop commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +372,161 @@ type KeylessRef struct {
+
+|                          | `Conditions.TrustedResourcesVerified` | `Conditions.Succeeded` |
+|--------------------------|---------------------------------------|------------------------|
+| all enforce policies pass| True                                  |                        |
```
is this set to true when all policies pass, not just enforce policies?
```suggestion
| all policies pass| True                                  |                        |
```

Yongxuanzhang replied:
Oh I think you're correct!

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,161 @@ type KeylessRef struct {
+
+**Examples:**
+* No Matching Policies:
+    * `trusted-resources-verification-no-match-policy` is set to `ignore`, then no updates on conditions, succeeded `conditions` are:
+      ```yaml
+      status:
+        conditions:
+        - lastTransitionTime: "2023-03-01T18:17:10Z"
+          message: All Steps have completed executing
+          reason: Succeeded
+          status: "True"
+          type: Succeeded
+      ```
````
in this case, we don't actually do anything to `Conditions.Succeeded` (there could be another failure, can't assume that it'd be "True")

Yongxuanzhang replied:
removed all unnecessary conditions to avoid confusions

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,161 @@ type KeylessRef struct {
+          status: "True"
+          type: Succeeded
+      ```
+    * `trusted-resources-verification-no-match-policy` is set to `warn`, only add `false` `ConditionTrustedResourcesVerified` to `conditions`:
+      ```yaml
+      status:
+        conditions:
+        - lastTransitionTime: "2023-03-01T18:17:05Z"
+          message: Trusted resource verification failed
+          reason: ResourceVerificationFailed
+          status: "False"
+          type: TrustedResourcesVerified
+        - lastTransitionTime: "2023-03-01T18:17:10Z"
+          message: All Steps have completed executing
+          reason: Succeeded
+          status: "True"
+          type: Succeeded
+      ```
````
in this case, we don't actually do anything to `Conditions.Succeeded` (there could be another failure, can't assume that it'd be "True")

jerop replied:
```suggestion
    * `trusted-resources-verification-no-match-policy` is set to `warn`, only add `false` `ConditionTrustedResourcesVerified` to `conditions`:
      ```yaml
      status:
        conditions:
        - lastTransitionTime: "2023-03-01T18:17:05Z"
          message: Trusted resource verification failed
          reason: ResourceVerificationFailed
          status: "False"
          type: TrustedResourcesVerified
      ```
```

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,161 @@ type KeylessRef struct {
+          type: Succeeded
+      ```
+* Matching Policies (no matter what `trusted-resources-verification-no-match-policy` value is):
+  * all policies pass, add `true` `ConditionTrustedResourcesVerified`:
+      ```yaml
+      status:
+        conditions:
+        - lastTransitionTime: "2023-03-01T18:17:05Z"
+          message: Trusted resource verification passed
+          reason: ResourceVerificationSucceeded
+          status: "True"
+          type: TrustedResourcesVerified
+        - lastTransitionTime: "2023-03-01T18:17:10Z"
+          message: All Steps have completed executing
+          reason: Succeeded
+          status: "True"
+          type: Succeeded
+      ```
````
in this case, we don't actually do anything to `Conditions.Succeeded` (there could be another failure, can't assume that it'd be "True")

```suggestion
  * all policies pass, add `true` `ConditionTrustedResourcesVerified`:
      ```yaml
      status:
        conditions:
        - lastTransitionTime: "2023-03-01T18:17:05Z"
          message: Trusted resource verification passed
          reason: ResourceVerificationSucceeded
          status: "True"
          type: TrustedResourcesVerified
      ```
```

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,161 @@ type KeylessRef struct {
+          status: "False"
+          type: Succeeded
+      ```
+     * only warn policies fail, then only add `false` `ConditionTrustedResourcesVerified`:
+      ```yaml
+      status:
+        conditions:
+        - lastTransitionTime: "2023-03-01T18:17:05Z"
+          message: Trusted resource verification failed
+          reason: ResourceVerificationFailed
+          status: "False"
+          type: TrustedResourcesVerified
+        - lastTransitionTime: "2023-03-01T18:17:10Z"
+          message: All Steps have completed executing
+          reason: Succeeded
+          status: "True"
+          type: Succeeded
+      ```
````
in this case, we don't actually do anything to `Conditions.Succeeded` (there could be another failure, can't assume that it'd be "True")

```suggestion
     * only warn policies fail, then only add `false` `ConditionTrustedResourcesVerified`:
      ```yaml
      status:
        conditions:
        - lastTransitionTime: "2023-03-01T18:17:05Z"
          message: Trusted resource verification failed
          reason: ResourceVerificationFailed
          status: "False"
          type: TrustedResourcesVerified
      ```
```

jerop commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +372,161 @@ type KeylessRef struct {
+| only warn policies fail  | False                                 |                        |
+
+**Examples:**
+* No Matching Policies:
```
can move these examples section under the "No Matching Policies" section above

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,161 @@ type KeylessRef struct {
+          status: "False"
+          type: Succeeded
+      ```
+* Matching Policies (no matter what `trusted-resources-verification-no-match-policy` value is):
````
can move these examples section under the "Matching Policies" section above

jerop commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+        status: "False"
+        type: TrustedResourcesVerified
+      - lastTransitionTime: "2023-03-01T18:17:10Z"
+        message: resource verification failed
```
```suggestion
        message: Trusted resource verification failed
```

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+          status: "True"
+          type: TrustedResourcesVerified
+      ```
+    * any enforce policy fails, add `false` `ConditionTrustedResourcesVerified` and return error in reconciler to set `Conditions.Succeeded` to `false`:
````
spacing/bulleting is off
```suggestion
  * any enforce policy fails, add `false` `ConditionTrustedResourcesVerified` and return error in reconciler to set `Conditions.Succeeded` to `false`:
```

jerop commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+          status: "False"
+          type: Succeeded
+      ```
+     * only warn policies fail, then only add `false` `ConditionTrustedResourcesVerified`:
````
spacing/bulleting is off

```suggestion
  * only warn policies fail, then only add `false` `ConditionTrustedResourcesVerified`:
```

jerop reviewed (approved):
Thanks for the addressing my feedback, the conditions are much clearer now 🎉

jerop commented:
@afrittoli @bobcatfish @jagathprakash @wlynch please take another look 🙏🏾

jerop commented:
once all assignees have approved, feel free to unhold (just don't want it to merge accidentally)

wlynch reviewed (commented):
/lgtm

bobcatfish commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+ * `warn`: Don't fail the taskrun/pipelinerun and log a warning if no matching policies are found.
+ * `fail`: Fail the taskrun/pipelinerun if no matching policies are found.
+
+**Note:** The current proposed `trusted-resources-verification-no-match-policy` will be added to replace the old `resource-verification-mode` in one release and this is not a backwards compatible change.
```
I think this is already on your radar, but please send a notification to tekton-dev@ and tekton-users@ to warn folks this is going to happen within one release, with some warning (at least a week I'm thinking?) before the release goes out 🙏

bobcatfish commented on teps/0091-trusted-resources.md:
```diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+
+### Condition Update
+
+Trusted resources should update the taskrun/pipelinerun’s condition to indicate if it passes verification or not. This can be done via knative api. Other options are listed in [alternatives](#alternatives).
```
nit: 'via the knative api' sounds like we're including some knative specific functionality our API. afaik [this is a kubernetes specific convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties); knative pkg just provides us with libraries that make this easier. If I'm understanding correctly, it might be clearer here to refer to the k8s conventions instead of knative

bobcatfish commented on teps/0091-trusted-resources.md:
````diff
@@ -364,26 +372,137 @@ type KeylessRef struct {
+        reason: ResourceVerificationFailed
+        status: "False"
+        type: TrustedResourcesVerified
+    ```
````
thank you for adding all this additional detail 🙏

bobcatfish reviewed (commented):
Thanks for all the updates @Yongxuanzhang ! There's one last nit from me about referring to the "knative API" but nothing blocking

/approve

Yongxuanzhang commented:
/hold cancel

wlynch reviewed (commented):
/lgtm
